rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -350,76 +354,111 @@ def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution
     @classmethod
     def on_graph_executor_start(cls):
         cls.pool_size = Config().num_node_workers
-        cls.executor = ProcessPoolExecutor(
-            max_workers=cls.pool_size,
+        cls._init_node_executor_pool()
+        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
+
+    @classmethod
+    def _init_node_executor_pool(cls):
+        cls.executor = Pool(
+            processes=cls.pool_size,
             initializer=cls.on_node_executor_start,
         )
-        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
 
     @classmethod
-    def on_graph_execution(cls, graph_data: GraphExecution):
+    def on_graph_execution(cls, graph_data: GraphExecution, cancel: threading.Event):
         prefix = get_log_prefix(graph_data.graph_exec_id, "*")
         logger.warning(f"{prefix} Start graph execution")
 
+        finished = False
+
+        def cancel_handler():
+            while not cancel.is_set():
+                cancel.wait(1)
+            if finished:
+                return
+            cls.executor.terminate()
+            logger.info(
+                f"{prefix} Terminated graph execution {graph_data.graph_exec_id}"
+            )
+            cls._init_node_executor_pool()
+
+        cancel_thread = threading.Thread(target=cancel_handler)
+        cancel_thread.start()
+
         try:
             queue = ExecutionQueue[NodeExecution]()
             for node_exec in graph_data.start_node_execs:
                 queue.add(node_exec)
 
-            futures: dict[str, Future] = {}
+            running_executions: dict[str, AsyncResult] = {}
             while not queue.empty():
-                execution = queue.get()
+                if cancel.is_set():
+                    return
+
+                exec_data = queue.get()
 
                 # Avoid parallel execution of the same node.
-                fut = futures.get(execution.node_id)
-                if fut and not fut.done():
+                execution = running_executions.get(exec_data.node_id)
+                if execution and not execution.ready():
                     # TODO (performance improvement):
                     #   Wait for the completion of the same node execution is blocking.
                     #   To improve this we need a separate queue for each node.
                     #   Re-enqueueing the data back to the queue will disrupt the order.
-                    cls.wait_future(fut, timeout=None)
+                    execution.wait()
 
-                futures[execution.node_id] = cls.executor.submit(
-                    cls.on_node_execution, queue, execution
+                running_executions[exec_data.node_id] = cls.executor.apply_async(
+                    cls.on_node_execution,
+                    (queue, exec_data),
+                    callback=lambda _: running_executions.pop(exec_data.node_id),
```

Comment:
@Pwuts are you sure the lambda is working here?

Can you try to add a try-catch and log any exceptions on this lambda?

Reply:
It wasn't always working properly, but it wasn't due to exceptions in the lambda itself.

It seems like the `Pool` is mixing up the callbacks to its jobs, only calling the callback to the latest added job. I don't understand why this is happening, but 4829c17fdc9b6bdaacea1208da7f141b66ab995e fixed it.

<details>
<summary>Technical breakdown of the issue and how I found it</summary>

### Summary
The callback is stored correctly on the `AsyncResult` object returned by `Pool.apply_async(..)`. But when the job itself finishes and an other new job has been scheduled in the meantime, the callback of the new job is called instead of the callback that belongs to the finished job.

### Code
`on_graph_execution` with extra log statements for transparency:
```py
    def on_graph_execution(cls, graph_data: GraphExecution, cancel: threading.Event):
        prefix = get_log_prefix(graph_data.graph_exec_id, "*")
        cls.logger.info(f"{prefix} Start graph execution")

        finished = False

        def cancel_handler():
            while not cancel.is_set():
                cancel.wait(1)
            logger.debug(f"{prefix} cancel_handler called; finished = {finished}")
            if finished:
                return
            cls.executor.terminate()
            logger.info(
                f"{prefix} Terminated graph execution {graph_data.graph_exec_id}"
            )
            cls._init_node_executor_pool()

        cancel_thread = threading.Thread(target=cancel_handler)
        cancel_thread.start()

        try:
            queue = ExecutionQueue[NodeExecution]()
            for node_exec in graph_data.start_node_execs:
                queue.add(node_exec)

            running_executions: dict[str, AsyncResult] = {}
            while not queue.empty():
                if cancel.is_set():
                    return

                exec_data = queue.get()
                logger.debug(f"{prefix} ID of `exec_data`: {id(exec_data)}")

                # Avoid parallel execution of the same node.
                execution = running_executions.get(exec_data.node_id)
                if execution and not execution.ready():
                    # TODO (performance improvement):
                    #   Wait for the completion of the same node execution is blocking.
                    #   To improve this we need a separate queue for each node.
                    #   Re-enqueueing the data back to the queue will disrupt the order.
                    execution.wait()

                # def make_exec_callback(exec_data: NodeExecution):
                #     node_id = exec_data.node_exec_id

                def exec_callback(result):
                    logger.debug(f"{prefix} Execution of node {exec_data.node_id} finished; removing it from `running_executions`")
                    logger.debug(f"{prefix} ID of `exec_data`: {id(exec_data)}")
                    logger.debug(f"{prefix} ID of `exec_callback`: {id(exec_callback)}")
                    logger.debug(f"{prefix} Pre-remove `running_executions`: {running_executions}")
                    running_executions.pop(exec_data.node_id)
                    logger.debug(f"{prefix} Post-remove `running_executions`: {running_executions}")
                    # return exec_callback
                logger.debug(f"{prefix} ID of `exec_callback`: {id(exec_callback)}")

                logger.debug(f"{prefix} Running node {exec_data.node_id} execution {exec_data.node_exec_id}")
                running_executions[exec_data.node_id] = cls.executor.apply_async(
                    cls.on_node_execution,
                    (queue, exec_data),
                    callback=exec_callback,
                )
                logger.debug(f"{prefix} Post-add `running_executions`: {running_executions}")

                for node_id, running_execution in running_executions.items():
                    logger.debug(f"{prefix} Node {node_id} result callback ID: {id(running_execution._callback)}")

                # Avoid terminating graph execution when some nodes are still running.
                while queue.empty() and running_executions:
                    logger.info(f"{prefix} Queue empty; running executions: {running_executions}")
                    for node_eid, execution in list(running_executions.items()):
                        if cancel.is_set():
                            return

                        if not queue.empty():
                            logger.debug(f"{prefix} Queue no longer empty; breaking...")
                            break  # yield to parent loop to execute new queue items

                        logger.debug(f"{prefix} Waiting for execution {node_eid} with status {{ready: {execution.ready()}}}")
                        execution.wait(3)

            cls.logger.info(f"{prefix} Completed graph execution")
```

### Output

Scheduling the start nodes:
```console
> ID of `exec_data`: 140243795155024
> ID of `exec_callback`: 140243795191712
> Running node daa4d7d5-4cab-4917-90ad-671c795db7cc execution b563b317-d42b-4295-bc8d-4b18b24648e4
> Post-add `running_executions`: {'daa4d7d5-4cab-4917-90ad-671c795db7cc': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94f0a0>}
> Node daa4d7d5-4cab-4917-90ad-671c795db7cc result callback ID: 140243795191712
> ID of `exec_data`: 140243795155424
> ID of `exec_callback`: 140243795191856
> Running node e8d3ef7d-ed04-432d-b03b-85cb157431d6 execution 83ba6261-4e8e-4b28-9812-1825942cdfbe
> Post-add `running_executions`: {'daa4d7d5-4cab-4917-90ad-671c795db7cc': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94f0a0>, 'e8d3ef7d-ed04-432d-b03b-85cb157431d6': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94ead0>}
> Node daa4d7d5-4cab-4917-90ad-671c795db7cc result callback ID: 140243795191712
> Node e8d3ef7d-ed04-432d-b03b-85cb157431d6 result callback ID: 140243795191856
```

Then the FIRST node execution job finishes:
```console
> [...|node-eid-b563b317-d42b-4295-bc8d-4b18b24648e4|-] Finished node execution
> Execution of node e8d3ef7d-ed04-432d-b03b-85cb157431d6 finished; removing it from `running_executions`
> ID of `exec_data`: 140243795155424
> ID of `exec_callback`: 140243795191856
> Pre-remove `running_executions`: {'daa4d7d5-4cab-4917-90ad-671c795db7cc': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94f0a0>, 'e8d3ef7d-ed04-432d-b03b-85cb157431d6': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94ead0>}
> Post-remove `running_executions`: {'daa4d7d5-4cab-4917-90ad-671c795db7cc': <multiprocessing.pool.ApplyResult object at 0x7f8d0d94f0a0>}
```
This is NOT the callback of the first node exec job, but of the second job!!!
</details>